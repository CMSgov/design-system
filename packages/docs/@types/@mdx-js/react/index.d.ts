declare module '@mdx-js/react' {
  const MDXProvider: React.ComponentType;
  export { MDXProvider };
}
