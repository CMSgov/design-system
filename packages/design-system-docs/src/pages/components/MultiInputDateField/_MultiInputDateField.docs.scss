/*
Date field (multi-input)

@uswds https://designsystem.digital.gov/components/memorable-date/

Date field (multi-input) provides date entry, spaced over three separate inputs for month, day, and year.

Markup: multiinputdatefield.example.html

Style guide: components.multi-input-date-field
*/

/*
`<MultiInputDateField>`

@react-example MultiInputDateField.example.jsx

@react-props MultiInputDateField.tsx

Style guide: components.multi-input-date-field.react
*/

/*
---

### When to use

- Use this date field when user flow requires inputs be separated. Generally used for memorable dates like date of birth, marriage date, or income start date.

### When to consider alternatives

- Don't use this date field if users are entering a date within the last year or a future date.

### Usage

- Allow users to enter one-digit numbers with or without the "0" prior to the number. Example: "1" as well as "01" for a month or day.
- Be sure each field is properly labeled â€” some countries enter dates in day, month, year order.
- It may be tempting to switch all or some of these text fields to drop downs, but these tend to be more difficult to use than text boxes.
- When handling date errors, if the part of the date that has an error is known, only highlight the part with an error. For example, if month and day are valid inputs, but the year is invalid, use `yearInvalid=true` so that only the year field will be highlighted red.
- If the entire date is incorrect, all fields can be highlighted. For example, if a date in the past is required and a date in the future is entered, all fields should be highlighted.

### Accessibility

- These text fields should follow the [accessibility guidelines for all text inputs]({{root}}/components/text-field/#guidance).
- Do not use JavaScript to auto advance the focus from one field to the next. This makes it difficult for keyboard-only users to navigate and correct mistakes.

### Related patterns

- [Date field (single-input)]({{root}}/components/single-input-date-field/)
- [Date field (with calendar)]({{root}}/components/date-field-with-calendar-picker/)
- [Text field]({{root}}/components/text-field/)
- [Grid]({{root}}/utilities/grid/)

Style guide: components.multi-input-date-field.guidance
*/
