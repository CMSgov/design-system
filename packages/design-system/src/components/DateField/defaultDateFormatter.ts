export interface DateObject {
  day: string;
  month: string;
  year: string;
}
// Prevents day/month greater than 2 digits and year greater than 4 digits
const standardLengthFormatter = ({ day, month, year }: DateObject): DateObject => ({
  day: day.length > 2 ? day.substring(0, 2) : day,
  month: month.length > 2 ? month.substring(0, 2) : month,
  year: year.length > 4 ? year.substring(0, 4) : year,
});

export const defaultDateFormatter = (dateObject: DateObject): DateObject =>
  standardLengthFormatter(dateObject);

export default defaultDateFormatter;
