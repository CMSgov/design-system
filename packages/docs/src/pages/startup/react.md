---
title: Using React components
weight: 3
---

The examples below assume you've installed the design system using NPM and have already setup your build system.

<h4 class="ds-h4 ds-u-font-size--base">Default imports</h4>

Individual components can be imported from their individual export file.

```jsx
import Button from '@cmsgov/design-system-core/dist/components/Button/Button';
```

<h4 class="ds-h4 ds-u-font-size--base">Named imports</h4>

Components can also be imported using the shorter syntax below.

<div class="ds-c-alert ds-c-alert--warn ds-u-margin-bottom--2 ds-u-font-size--small">
  <div class="ds-c-alert__body">
    <h3 class="ds-c-alert__heading ds-u-font-size--base">Performance note</h3>
    <p class="ds-c-alert__text">
      This approach may result in a much larger file than you intend. Depending on what module bundler you use, all of the design system's React components may be included in the bundled file even if you didn't specifically import them. This can be avoided by enabling features like [tree shaking in Webpack](https://webpack.js.org/guides/tree-shaking/).
    </p>
  </div>
</div>

```jsx
import { Button } from '@cmsgov/design-system-core';
```

[View an example](https://github.com/CMSgov/design-system/blob/master/examples/react-app/src/scripts/index.js)

<h2 id="examples" class="ds-h2 ds-u-color--primary-darker">Examples</h2>

Additional examples of the design system in use can be viewed on GitHub. These projects demonstrate the various ways you can incorporate the design system into your development process and various use cases.

<a href="https://github.com/CMSgov/design-system/tree/master/examples/" class="ds-c-button">Browse example projects</a>

<h2 id="need-help" class="ds-h2 ds-u-color--primary-darker">Need help or ran into an issue?</h2>

If you're having trouble installing or setting up the design system, or if you think you've found a bug, feel free to [open an issue on GitHub](https://github.com/CMSgov/design-system/issues).
