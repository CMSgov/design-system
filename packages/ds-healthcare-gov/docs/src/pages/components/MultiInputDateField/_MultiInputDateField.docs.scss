/*
---

### When to use

- Don't use this date field.

### Related patterns

- [Date field (single-input)]({{root}}/components/single-input-date-field/)
- [Date field (with calendar)]({{root}}/components/date-picker/)
- [Text field]({{root}}/components/text-field/)
- [Grid]({{root}}/utilities/grid/)

Style guide: components.multi-input-date-field.guidance
*/
