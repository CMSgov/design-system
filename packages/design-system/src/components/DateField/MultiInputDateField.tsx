import React from 'react';
import DateInput from './DateInput';
import defaultDateFormatter from './defaultDateFormatter';
import { FormFieldProps, useFormLabel } from '../FormLabel';
import { Label } from '../Label';
import { t } from '../i18n';
import useId from '../utilities/useId';

export type DateFieldDayDefaultValue = string | number;
export type DateFieldDayValue = string | number;
export type DateFieldMonthDefaultValue = string | number;
export type DateFieldMonthValue = string | number;
export type DateFieldYearDefaultValue = string | number;
export type DateFieldYearValue = string | number;

export interface DateFieldProps extends Omit<FormFieldProps, 'label'> {
  /**
   * Adds `autocomplete` attributes `bday-day`, `bday-month` and `bday-year` to the corresponding `<MultiInputDateField>` inputs
   */
  autoComplete?: boolean;
  /**
   * Optional method to format the `input` field values. If this
   * method is provided, the returned value will be passed as a second argument
   * to the `onBlur` and `onChange` callbacks. This method receives an object as
   * its only argument, in the shape of: `{ day, month, year }`
   *
   * By default `dateFormatter` will be set to the `defaultDateFormatter` function, which prevents days/months more than 2 digits & years more than 4 digits.
   */
  dateFormatter?: (...args: any[]) => any;
  /**
   * The primary label, rendered above the individual month/day/year fields
   */
  label?: React.ReactNode;
  /**
   * A unique ID to be used for the MultiInputDateField label. If one isn't provided, a unique ID will be generated.
   */
  labelId?: string;
  /**
   * Text showing the requirement ("Required", "Optional", etc.). See [Required and Optional Fields]({{root}}/guidelines/forms/#required-and-optional-fields).
   */
  requirementLabel?: React.ReactNode;
  /**
   * Called anytime any date input is blurred
   */
  onBlur?: (...args: any[]) => any;
  /**
   * Called when any date input is blurred and the focus does not land on one
   * of the other date inputs inside this component (i.e., when the whole
   * component loses focus)
   */
  onComponentBlur?: (...args: any[]) => any;
  /**
   * Called anytime any date input is changed
   */
  onChange?: (...args: any[]) => any;
  /**
   * Label for the day field
   */
  dayLabel?: React.ReactNode;
  /**
   * `name` for the day `input` field
   */
  dayName?: string;
  /**
   * Initial value for the day `input` field. Use this for an uncontrolled
   * component; otherwise, use the `dayValue` property.
   */
  dayDefaultValue?: DateFieldDayDefaultValue;
  /**
   * Access a reference to the day `input`
   */
  dayFieldRef?: (...args: any[]) => any;
  /**
   * Apply error styling to the day `input`
   */
  dayInvalid?: boolean;
  /**
   * Sets the day input's `value`. Use this in combination with `onChange`
   * for a controlled component; otherwise, set `dayDefaultValue`.
   */
  dayValue?: DateFieldDayValue;
  /**
   * Label for the month field
   */
  monthLabel?: React.ReactNode;
  /**
   * `name` for the month `input` field
   */
  monthName?: string;
  /**
   * Initial value for the month `input` field. Use this for an uncontrolled
   * component; otherwise, use the `monthValue` property.
   */
  monthDefaultValue?: DateFieldMonthDefaultValue;
  /**
   * Access a reference to the month `input`
   */
  monthFieldRef?: (...args: any[]) => any;
  /**
   * Apply error styling to the month `input`
   */
  monthInvalid?: boolean;
  /**
   * Sets the month input's `value`. Use this in combination with `onChange`
   * for a controlled component; otherwise, set `monthDefaultValue`.
   */
  monthValue?: DateFieldMonthValue;
  /**
   * Initial value for the year `input` field. Use this for an uncontrolled
   * component; otherwise, use the `yearValue` property.
   */
  yearDefaultValue?: DateFieldYearDefaultValue;
  /**
   * Access a reference to the year `input`
   */
  yearFieldRef?: (...args: any[]) => any;
  /**
   * Apply error styling to the year `input`
   */
  yearInvalid?: boolean;
  /**
   * Label for the year `input` field
   */
  yearLabel?: React.ReactNode;
  /**
   * `name` for the year field
   */
  yearName?: string;
  /**
   * Sets the year input's `value`. Use this in combination with `onChange`
   * for a controlled component; otherwise, set `yearDefaultValue`.
   */
  yearValue?: DateFieldYearValue;
}

/**
 * For information about how and when to use this component,
 * [refer to its full documentation page](https://design.cms.gov/components/date-field/multi-input-date-field/).
 */
export function MultiInputDateField(props: DateFieldProps): React.ReactElement {
  const id = useId('date-field--', props.id);
  const { labelProps, fieldProps, wrapperProps, bottomError } = useFormLabel({
    label: t('dateField.label'),
    hint: t('dateField.hint'),
    dayName: 'day',
    monthName: 'month',
    yearName: 'year',
    dateFormatter: defaultDateFormatter,
    ...props,
    labelComponent: 'legend',
    wrapperIsFieldset: true,
    id,
  });

  delete fieldProps.errorId;

  return (
    <fieldset {...wrapperProps}>
      <Label {...labelProps} />
      <DateInput {...fieldProps} />
      {bottomError}
    </fieldset>
  );
}

export default MultiInputDateField;
