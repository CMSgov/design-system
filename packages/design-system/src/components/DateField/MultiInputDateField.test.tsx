import React from 'react';
jest.mock('lodash/uniqueId', () => (str) => `${str}snapshot`);
import { MultiInputDateField } from './MultiInputDateField';
import defaultDateFormatter from './defaultDateFormatter';
import { render, screen } from '@testing-library/react';

describe('MultiInputDateField', () => {
  it('renders', () => {
    const { container } = render(<MultiInputDateField />);

    expect(container.querySelectorAll('label')).toHaveLength(3);
    expect(screen.getAllByText('/')).toHaveLength(2);

    const monthInput = screen.getByLabelText(/month/i);
    expect(monthInput).toBeInTheDocument();
    expect(monthInput).toHaveAttribute('maxlength', '2');
    expect(monthInput.classList).toContain('ds-c-field--month');

    const dayInput = screen.getByLabelText(/day/i);
    expect(dayInput).toBeInTheDocument();
    expect(dayInput).toHaveAttribute('maxlength', '2');
    expect(dayInput.classList).toContain('ds-c-field--day');

    const yearInput = screen.getByLabelText(/year/i);
    expect(yearInput).toBeInTheDocument();
    expect(yearInput).toHaveAttribute('maxlength', '4');
    expect(yearInput.classList).toContain('ds-c-field--year');

    const inputs = screen.getAllByRole('textbox');
    expect(inputs).toHaveLength(3);
    inputs.forEach((input) => {
      expect(input).toHaveAttribute('aria-invalid', 'false');
      expect(input).toHaveAttribute('inputmode', 'numeric');
      expect(input).toHaveAttribute('pattern', '[0-9]*');
      expect(input).toHaveAttribute('type', 'text');
    });
  });

  describe('defaultDateFormatter', () => {
    it('prevents day over 2 digits', () => {
      const date = { day: '100', month: '10', year: '2000' };
      const formattedDate = defaultDateFormatter(date);
      expect(formattedDate.day).toBe('10');
      expect(formattedDate.month).toBe(date.month);
      expect(formattedDate.year).toBe(date.year);
    });

    it('prevents month over 2 digits', () => {
      const date = { day: '10', month: '100', year: '2000' };
      const formattedDate = defaultDateFormatter(date);
      expect(formattedDate.day).toBe(date.day);
      expect(formattedDate.month).toBe('10');
      expect(formattedDate.year).toBe(date.year);
    });

    it('prevents year over 4 digits', () => {
      const date = { day: '10', month: '10', year: '20000' };
      const formattedDate = defaultDateFormatter(date);
      expect(formattedDate.day).toBe(date.day);
      expect(formattedDate.month).toBe(date.month);
      expect(formattedDate.year).toBe('2000');
    });
  });
});
