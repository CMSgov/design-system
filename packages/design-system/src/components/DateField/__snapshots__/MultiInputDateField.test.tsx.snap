// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiInputDateField renders 1`] = `
<DocumentFragment>
  <fieldset
    aria-describedby="field_snapshot-hint"
    aria-invalid="false"
    class="ds-c-fieldset"
  >
    <legend
      class="ds-c-label"
      id="field_snapshot-label"
    >
      Date
    </legend>
    <div
      class="ds-c-field__hint"
      id="field_snapshot-hint"
    >
      For example: 4 / 28 / 1986
    </div>
    <div
      class="ds-c-datefield__container ds-l-form-row"
    >
      <div
        class="ds-u-clearfix ds-l-col--auto"
      >
        <label
          class="ds-c-label ds-c-datefield__label"
          for="field_snapshot"
          id="field_snapshot-label"
        >
          Month
        </label>
        <input
          aria-invalid="false"
          class="ds-c-field ds-c-field--month"
          id="field_snapshot"
          inputmode="numeric"
          maxlength="2"
          name="month"
          pattern="[0-9]*"
          type="text"
          value=""
        />
      </div>
      <span
        class="ds-c-datefield__separator"
      >
        /
      </span>
      <div
        class="ds-u-clearfix ds-l-col--auto"
      >
        <label
          class="ds-c-label ds-c-datefield__label"
          for="field_snapshot"
          id="field_snapshot-label"
        >
          Day
        </label>
        <input
          aria-invalid="false"
          class="ds-c-field ds-c-field--day"
          id="field_snapshot"
          inputmode="numeric"
          maxlength="2"
          name="day"
          pattern="[0-9]*"
          type="text"
          value=""
        />
      </div>
      <span
        class="ds-c-datefield__separator"
      >
        /
      </span>
      <div
        class="ds-u-clearfix ds-l-col--auto"
      >
        <label
          class="ds-c-label ds-c-datefield__label"
          for="field_snapshot"
          id="field_snapshot-label"
        >
          Year
        </label>
        <input
          aria-invalid="false"
          class="ds-c-field ds-c-field--year"
          id="field_snapshot"
          inputmode="numeric"
          maxlength="4"
          name="year"
          pattern="[0-9]*"
          type="text"
          value=""
        />
      </div>
    </div>
  </fieldset>
</DocumentFragment>
`;
